### PR TITLE
Add dictionary data structure

### LINKED ISSUE #50: Add a dictionary data structure
Inspired by the discussion at https://github.com/LLK/scratch-vm/issues/1030#issuecomment-379405892.

I will start by tinkering around to see how this can be implemented.

### PR DESCRIPTION
**Pull request by tansly**
Closes #50.

TODO:
- [x]  Add duplication logic for dictionaries here: https://github.com/FBDY/bb-vm/blob/48f349a89e0961be6977ff1af63fa40c56dfc1cd/src/engine/target.js#L404-L408
For now, they fall to the `else` case which makes a plain assignment. But with a JS object (which is what our dictionary representation is) I guess this will not produce correct behaviour.
- [x] Weird add/get bug -> fixed by b9a82b2
- [x] Show/hide blocks not working -> fixed by b9a82b2
- [x] Add execution support for new blocks as they are added in https://github.com/FBDY/bb-blocks/pull/22
- [x] Design and implement a good representation for the dictionary monitor in GUI https://github.com/FBDY/bb-gui/pull/135
- [x] There are problems with the monitor not updating when the dictionary is cleared by the `delete all of dict` block. However, this will be resolved as a side effect when the monitor work is completed because of reasons that I'm too lazy to explain here. But in any case, don't forget to check if it works when this work is finalized.
- [x] Provide a way to iterate over keys in a dictionary. A `for key in` kind of block?

### CODE REVIEW

**Comment by tansly** [comment]
In order to decide on the internal representaion of dictionaries, I was reading some answers from https://stackoverflow.com/questions/7196212/how-to-create-dictionary-and-add-key-value-pairs-dynamically.

I decided to go with plain JS objects. Starting now.

**Comment by tansly** [comment]
I've added the code for gettling dictionary contents. It is not complete, though:
Looks like I need to delve into the GUI so that the React monitor shows the dictionary in a pretty format. I'm not exactly sure what to do. Will examine in the following days.

**Comment by tansly** [comment]
Not related to the GUI stuff, just want to note here for my later self: Don't forget to take a look at `getAllVariableAndListReferences()` later.

File: src/blocks/scratch3_data.js [hunk 1]
```diff
@@ -252,6 +253,34 @@ class Scratch3DataBlocks {
     static get LIST_ITEM_LIMIT () {
         return 200000;
     }
+
+    /*
+     * TODO: In the GUI, dictionary contents does not have a pretty title;
+     * it shows data_dictcontents (the opcode). Fix it.
+     * TODO: Find out how the GUI displays lists the way it does and if possible
```
Comment (tansly) [inline src/blocks/scratch3_data.js]: I'll try to sort this out after implementing `data_addtodict` operation.

Comment (tansly) [reply src/blocks/scratch3_data.js]: After implementing `data_itemofdict` as well.

**Comment by tansly** [comment]
Add/get working! So far so good...

File: src/blocks/scratch3_data.js (hunk 1 above)
Comment (tansly) [reply src/blocks/scratch3_data.js]: Add/get working -- cool and good.

Let's try to sort the GUI stuff...

**Comment by tansly** [comment]
Noticed a very strange bug. When I:
* Create multiple dictionaries
* Add an item to one of them

It seems like the item is added to all of the dictionaries, or something else is going on that makes it seem like so.

**Comment by tansly** [comment]
The show/hide blocks do not work right now, but too many things are broken right now so I believe it will start working eventually.

**Comment by tansly** [comment]
> Noticed a very strange bug. When I:
> 
>     * Create multiple dictionaries
> 
>     * Add an item to one of them
> 
> 
> It seems like the item is added to all of the dictionaries, or something else is going on that makes it seem like so.

Dictionary variables that I create have `undefined` as their `name` property:

`Object { id: "]bs-MYS-rl*3MRhGB}%=", name: undefined, type: "dict", isCloud: false, value: {} }`

I don't think this is the root cause, but it may be a symptom that will lead me to it.

**Comment by tansly** [comment]
I'll be maintaining a TODO list in the first comment in order to make this discussion more manageable. If anyone at @FBDY/team wants to help, feel free to attack these items with prior notice.

**Comment by tansly** [comment]
> > Noticed a very strange bug. When I:
> > ```
> > * Create multiple dictionaries
> > 
> > * Add an item to one of them
> > ```
> > It seems like the item is added to all of the dictionaries, or something else is going on that makes it seem like so.
> 
> Dictionary variables that I create have `undefined` as their `name` property:
> 
> `Object { id: "]bs-MYS-rl*3MRhGB}%=", name: undefined, type: "dict", isCloud: false, value: {} }`
> 
> I don't think this is the root cause, but it may be a symptom that will lead me to it.


b9a82b2 fixes this problem :rocket:

**Comment by tansly** [comment]
> The show/hide blocks do not work right now, but too many things are broken right now so I believe it will start working eventually.

This is fixed by b9a82b2 as well.

**Comment by tansly** [comment]
This is probably ready to merge

File: src/blocks/scratch3_data.js [hunk 2]
```diff
@@ -252,6 +262,106 @@ class Scratch3DataBlocks {
...
+        const dict = util.target.lookupOrCreateDict(
+            args.DICT.id, args.DICT.name);
+
+        if (util.thread.updateMonitor) {
+            // Return original array representation if up-to-date, which doesn't trigger monitor update.
+            if (dict._monitorUpToDate) return dict.arrayRepr;
+            // If value changed, reset the flag, update the stored array and return a copy to trigger monitor update.
+            // Because monitors use Immutable data structures, only new objects trigger updates.
+            dict._monitorUpToDate = true;
+            // TODO: Make this prettier together with the help of bb-gui
```
Comment (tansly) [inline src/blocks/scratch3_data.js]: Remove this comment

File: src/virtual-machine.js [hunk 3]
```diff
@@ -1509,6 +1509,12 @@ class VirtualMachine extends EventEmitter {
             const variable = target.lookupVariableById(variableId);
             if (variable) {
                 variable.value = value;
+                // XXX: Temporary hack
```
Comment (tansly) [inline src/virtual-machine.js:1515]: Not sure if this is necessary. Don't forget to check after all this variable API migration stuff is over.

Comment (tansly) [reply src/virtual-machine.js:1515]: > Not sure if this is necessary. Don't forget to check after all this variable API migration stuff is over.

It is actually necessary, forget about it.

**Comment by filips123** [comment]
@tansly Given that dictionary support could be useful and that this is fork based on Scratch, could you do something similar for original Scratch?

**Comment by tansly** [comment]
> @tansly Given that dictionary support could be useful and that this is fork based on Scratch, could you do something similar for original Scratch?

@filips123 The original plan was to open a PR on Scratch for this, but sadly we could not find the time to get this to an acceptable state. We were on a tight schedule to get this working so it contains some invasive changes in the codebase and also some dirty hacks. I would love to clean this up and contribute it to Scratch, but sadly I can't find the time to deal with it these days.

**Pull request merged**